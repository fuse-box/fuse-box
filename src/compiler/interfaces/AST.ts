import { IASTScope } from '../Visitor/Visitor';

export interface ASTNode {
  raw?: string;
  type: keyof typeof ASTType;

  $assign_pattern?: boolean;
  $fuse_classInitializers?: Array<ASTNode>;
  $fuse_class_declaration_visited?: boolean;
  $fuse_decoratorForClassIdentifier?: string;
  $fuse_decoratorsCaptured?: boolean;
  $fuse_decoratorsProcessed?: boolean;
  $fuse_visited?: boolean;
  $transpiled?: boolean;
  alternate?: ASTNode;
  base?: ASTNode;
  chain?: ASTNode;
  consequent?: ASTNode;
  decorators?: Array<ASTNode>;
  definite?: boolean;
  elementTypes?: Array<ASTNode>;
  elements?: Array<ASTNode>;
  expressions?: Array<ASTNode>;
  initializer?: ASTNode;
  literal?: ASTNode;
  moduleReference?: ASTNode;
  optional?: boolean;
  prefix?: boolean;
  returnType?: ASTNode;
  scope?: IASTScope;
  shorthand?: boolean;
  source?: ASTNode;
  sourceType?: string;
  specifiers?: Array<ASTNode>;
  test?: ASTNode;
  typeAnnotation?: ASTNode;
  typeName?: ASTNode;
  loc?: { end: { column: number; line: number }; start: { column: number; line: number } };

  imported?: ASTNode;
  members?: Array<ASTNode>;
  method?: boolean;

  JSXElement?: ASTNode;

  abstract?: boolean;
  accessibility?: string;
  argument?: ASTNode;
  arguments?: Array<ASTNode>;
  async?: boolean;
  attributes?: Array<ASTNode>;
  body?: ASTNode | Array<ASTNode>;
  callee?: ASTNode;
  children?: Array<ASTNode>;
  closingElement?: ASTNode;
  computed?: boolean;
  declaration?: ASTNode;
  declarations?: Array<ASTNode>;
  declare?: boolean;
  exported?: ASTNode;
  expression?: any;
  generator?: boolean;
  id?: ASTNode;
  implements?: Array<ASTNode>;
  init?: null | ASTNode;
  key?: ASTNode;
  kind?: string;
  left?: ASTNode;
  local?: ASTNode;
  name?: any;
  object?: ASTNode;
  openingElement?: ASTNode;
  operator?: string;
  parameter?: ASTNode;
  params?: Array<ASTNode>;
  properties?: Array<ASTNode>;
  property?: ASTNode;
  quasi?: ASTNode;
  quasis?: Array<ASTNode>;
  right?: ASTNode;
  static?: boolean;
  superClass?: ASTNode;
  tag?: ASTNode;
  tail?: ASTNode;
  value?: any;
}

export const ASTType = {
  EnumDeclaration: 'TSEnumDeclaration',
  EnumMember: 'TSEnumMember',

  ParameterProperty: 'TSParameterProperty',
  StringKeyword: 'TSStringKeyword',
  TypeAnnotation: 'TSTypeAnnotation',

  AbstractClassProperty: 'TSAbstractClassProperty',
  AbstractMethodDefinition: 'TSAbstractMethodDefinition',
  AsExpression: 'TSAsExpression',
  DeclareFunction: 'TSDeclareFunction',
  DeclareKeyword: 'TSDeclareKeyword',
  InterfaceDeclaration: 'TSInterfaceDeclaration',
  NonNullExpression: 'TSNonNullExpression',
  TypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TypeAssertion: 'TSTypeAssertion',

  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  Block: 'Block',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  HTMLClose: 'HTMLClose',
  HTMLOpen: 'HTMLOpen',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Line: 'Line',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  OptionalCallExpression: 'OptionalCallExpression',
  OptionalMemberExpression: 'OptionalMemberExpression',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  module: 'module',
  script: 'script',

  AbstractKeyword: 'AbstractKeyword',

  AnyKeyword: 'TSAnyKeyword',
  ArrayType: 'TSArrayType',

  AsyncKeyword: 'AsyncKeyword',
  BigIntKeyword: 'BigIntKeyword',
  BooleanKeyword: 'TSBooleanKeyword',
  CallSignatureDeclaration: 'CallSignatureDeclaration',
  ClassImplements: 'ClassImplements',
  ConditionalType: 'ConditionalType',
  ConstructSignatureDeclaration: 'ConstructSignatureDeclaration',
  ConstructorType: 'ConstructorType',
  EmptyBodyFunctionExpression: 'EmptyBodyFunctionExpression',
  ExportAssignment: 'ExportAssignment',

  ExportKeyword: 'ExportKeyword',
  ExternalModuleReference: 'ExternalModuleReference',
  FunctionType: 'TSFunctionType',
  ImportEqualsDeclaration: 'TSImportEqualsDeclaration',
  ImportExpression: 'ImportExpression',
  ImportType: 'ImportType',
  IndexSignature: 'IndexSignature',
  IndexedAccessType: 'IndexedAccessType',
  InferType: 'InferType',

  InterfaceBody: 'InterfaceBody',
  InterfaceHeritage: 'InterfaceHeritage',
  IntersectionType: 'IntersectionType',
  LiteralType: 'TSLiteralType',
  MappedType: 'MappedType',
  MethodSignature: 'MethodSignature',
  ModuleBlock: 'ModuleBlock',
  ModuleDeclaration: 'TSModuleDeclaration',
  NamespaceExportDeclaration: 'NamespaceExportDeclaration',
  NeverKeyword: 'TSNeverKeyword',

  NullKeyword: 'TSNullKeyword',
  NumberKeyword: 'TSNumberKeyword',
  ObjectKeyword: 'ObjectKeyword',
  OptionalType: 'OptionalType',

  ParenthesizedType: 'ParenthesizedType',
  PrivateKeyword: 'PrivateKeyword',
  PropertySignature: 'PropertySignature',
  ProtectedKeyword: 'ProtectedKeyword',
  PublicKeyword: 'PublicKeyword',
  QualifiedName: 'QualifiedName',
  ReadonlyKeyword: 'ReadonlyKeyword',
  RestType: 'RestType',
  StaticKeyword: 'StaticKeyword',

  SymbolKeyword: 'SymbolKeyword',
  ThisType: 'ThisType',
  TupleType: 'TSTupleType',

  JSDocNullableType: 'JSDocNullableType',
  JSDocUnknownType: 'JSDocUnknownType',
  TypeLiteral: 'TypeLiteral',
  TypeOperator: 'TypeOperator',
  TypeParameter: 'TypeParameter',
  TypeParameterDeclaration: 'TypeParameterDeclaration',
  TypeParameterInstantiation: 'TypeParameterInstantiation',
  TypePredicate: 'TypePredicate',
  TypeQuery: 'TypeQuery',
  TypeReference: 'TSTypeReference',
  UnaryExpression: 'UnaryExpression',
  UndefinedKeyword: 'TSUndefinedKeyword',
  UnionType: 'UnionType',
  UnknownKeyword: 'UnknownKeyword',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  VoidKeyword: 'TSVoidKeyword',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
};
