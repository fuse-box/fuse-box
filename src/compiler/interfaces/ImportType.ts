export enum ImportType {
  REQUIRE,
  FROM,
  RAW_IMPORT,
  DYNAMIC,
}
