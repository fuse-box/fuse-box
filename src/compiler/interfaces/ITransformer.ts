import { ICompilerOptions } from '../../compilerOptions/interfaces';
import { Context } from '../../core/context';
import { IModule } from '../../moduleResolver/module';
import { IProductionContext } from '../../production/ProductionContext';
import { IVisit, IVisitorMod } from '../Visitor/Visitor';
import { ASTNode } from './AST';
import { ImportType } from './ImportType';

export type ITranformerCallback = (visit: IVisit) => undefined | IVisitorMod | void;

export interface ITransformerVisitors {
  onEachNode?: ITranformerCallback;
  onTopLevelTraverse?: ITranformerCallback;
}

export interface IRequireStatementModuleOptions {
  breakDependantsCache?: boolean;
}
export interface ITransformerCommon {
  compilerOptions: ICompilerOptions;
  ctx: Context;
  module: IModule;
  onRequireCallExpression?: (
    importType: ImportType,
    node: ASTNode,
    moduleOptions?: IRequireStatementModuleOptions,
  ) => void;
}

export interface ITransformerProduction {
  ctx: Context;
  module: IModule;
  productionContext: IProductionContext;
}

export interface ITransformer {
  target?: {
    test?: RegExp;
    type?: 'js' | 'js_ts' | 'ts';
  };
  commonVisitors?: (props?: ITransformerCommon) => ITransformerVisitors;
  productionWarmupPhase?: (props?: ITransformerProduction) => ITransformerVisitors;
}
