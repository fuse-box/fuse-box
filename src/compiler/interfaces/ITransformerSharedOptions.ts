import { ImportType } from './ImportType';
import { ASTNode } from './AST';

export interface ITransformerSharedOptions {
  onRequireCallExpression?: (importType: ImportType, node: ASTNode) => void;
}
