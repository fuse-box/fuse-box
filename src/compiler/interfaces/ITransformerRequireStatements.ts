import { ASTNode } from './AST';
import { IRequireStatementModuleOptions } from './ITransformer';
import { ImportType } from './ImportType';

export interface ITransformerRequireStatement {
  importType: ImportType;
  moduleOptions?: IRequireStatementModuleOptions;
  statement: ASTNode;
  value?: string;
}
export type ITransformerRequireStatementCollection = Array<ITransformerRequireStatement>;
