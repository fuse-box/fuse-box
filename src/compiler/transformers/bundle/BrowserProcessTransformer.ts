import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import {
  isPropertyOrPropertyAccess,
  createASTFromObject,
  defineVariable,
  createRequireStatement,
} from '../../Visitor/helpers';
import { ImportType } from '../../interfaces/ImportType';

interface BrowserProcessTransformProps {
  env?: { [key: string]: string };
}
export type IBrowserProcessTransform = BrowserProcessTransformProps & ITransformerSharedOptions;
export function BrowserProcessTransformer(options: IBrowserProcessTransform) {
  options = options || {};

  let globalEnvInserted = false;
  let entireEnvInserted = false;
  return (visit: IVisit): IVisitorMod => {
    const { node, parent } = visit;

    const accessList = isPropertyOrPropertyAccess(node, parent, 'process');
    if (accessList) {
      const variableAmount = accessList.length;
      if (variableAmount === 3 && accessList[1] === 'env') {
        const keyName = accessList[2];

        if (options.env && options.env[keyName] !== undefined) {
          return { replaceWith: { type: 'Literal', value: options.env[keyName].toString() } };
        }
        return { replaceWith: { type: 'Identifier', value: 'undefined' } };
      }
      if (variableAmount === 2) {
        switch (accessList[1]) {
          case 'version':
            return { replaceWith: { type: 'Literal', value: process.version }, avoidReVisit: true };
          case 'versions':
            return { replaceWith: { type: 'ObjectExpression', properties: [] }, avoidReVisit: true };
          case 'title':
            return { replaceWith: { type: 'Literal', value: 'browser' }, avoidReVisit: true };
          case 'umask':
            return { replaceWith: { type: 'Literal', value: 0 }, avoidReVisit: true };
          case 'browser':
            return { replaceWith: { type: 'Literal', value: true }, avoidReVisit: true };
          case 'cwd':
            return { replaceWith: { type: 'Literal', value: './' }, avoidReVisit: true };
          case 'env':
            if (options.env) {
              const response: IVisitorMod = {};
              if (!globalEnvInserted) {
                response.prependToBody = [defineVariable('___env', createASTFromObject(options.env))];
                globalEnvInserted = true;
              }
              response.replaceWith = { type: 'Identifier', name: '___env' };
              return response;
            } else return { replaceWith: createASTFromObject({}), avoidReVisit: true };

          default:
            // inserting require("process")
            // because we cannot match a know variable
            if (!entireEnvInserted) {
              entireEnvInserted = true;
              const statement = createRequireStatement('process', 'process');
              if (options.onRequireCallExpression)
                options.onRequireCallExpression(ImportType.REQUIRE, statement.reqStatement);
              return {
                avoidReVisit: true,
                prependToBody: [statement.statement],
              };
            }
            break;
        }
      }
    }
    return;
  };
}
