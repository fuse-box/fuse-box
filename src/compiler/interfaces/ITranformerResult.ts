import { ASTNode } from './AST';
import { ITransformerRequireStatementCollection } from './ITransformerRequireStatements';

export interface ITransformerResult {
  ast?: ASTNode;
  requireStatementCollection: ITransformerRequireStatementCollection;
}
