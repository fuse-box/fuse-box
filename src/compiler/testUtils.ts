import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ImportType } from './interfaces/ImportType';
import { ITransformer } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { parseTypeScript } from './parser';
import { createGlobalContext } from './program/GlobalContext';
import { transpileModule } from './program/transpileModule';
import { GlobalContextTransformer } from './transformers/GlobalContextTransformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });

  // const ast = buntis.parseTSModule(props.code, {
  //   directives: true,
  //   jsx: props.jsx,
  //   next: true,
  //   loc: true,
  // });
  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors({ ...props.props, onRequireCallExpression }));
    }
  }
  transpileModule({
    ast: ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
