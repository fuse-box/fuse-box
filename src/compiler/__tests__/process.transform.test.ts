import { ImportType } from '../interfaces/ImportType';

import { initCommonTransform } from '../testUtils';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';

const testTranspile = (props: { code: string; target?: string; env?: any }) => {
  return initCommonTransform({
    props: {
      ctx: { config: { env: props.env || {}, target: props.target || 'browser' } },
    },
    transformers: [BrowserProcessTransformer()],
    code: props.code,
  });
};

describe('Process transform test', () => {
  describe('process.env.***', () => {
    it('should replace process.env.NODE_ENV', () => {
      const result = testTranspile({
        target: 'browser',
        env: { NODE_ENV: 'development' },
        code: `
          console.log(process.env.NODE_ENV);
      `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should replace with undefined if value not found', () => {
      const result = testTranspile({
        target: 'browser',
        env: {},
        code: `
          console.log(process.env.NODE_ENV);
      `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Should not touch process', () => {
    it('should transform process.version', () => {
      const result = testTranspile({
        code: `
          var process = {};
          console.log(process.version);
      `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should inject process on assignment', () => {
      const result = testTranspile({
        code: `
          process.env.NODE_ENV = "hey";

      `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });
  describe('process.xxx', () => {
    it('should transform process.version', () => {
      const result = testTranspile({
        code: `
          console.log(process.version);
      `,
      });

      expect(result.code).toMatch(/console.log\("v\d+/);
    });

    it('should transform process.versions', () => {
      const result = testTranspile({
        code: `console.log(process.versions);`,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should transform process.title', () => {
      const result = testTranspile({
        code: `console.log(process.title);`,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should transform process.umask', () => {
      const result = testTranspile({
        code: `console.log(process.umask);`,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should transform process.browser', () => {
      const result = testTranspile({
        code: `console.log(process.browser);`,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should transform process.cwd', () => {
      const result = testTranspile({
        code: `console.log(process.cwd());`,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should transform process.env', () => {
      const result = testTranspile({
        code: `console.log(process.env);`,
        target: 'browser',
        env: { foo: 'bar' },
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should transform process.env (not add twice)', () => {
      const result = testTranspile({
        code: `
        alert(process.env)
        console.log(process.env);
        `,
        target: 'browser',
        env: { foo: 'bar' },
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should inject process if a variable is not recognized', () => {
      const result = testTranspile({
        code: `
        console.log(process.some());
        `,
      });

      expect(result.requireStatementCollection).toEqual([
        {
          importType: ImportType.REQUIRE,
          statement: {
            type: 'CallExpression',
            callee: { type: 'Identifier', name: 'require' },
            arguments: [{ type: 'Literal', value: 'process' }],
          },
        },
      ]);
      expect(result.code).toMatchSnapshot();
    });
  });
});
