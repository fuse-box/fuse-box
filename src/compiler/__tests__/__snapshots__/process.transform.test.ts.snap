// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process transform test process.env.*** should replace process.env.NODE_ENV 1`] = `
"console.log(\\"development\\");
"
`;

exports[`Process transform test process.env.*** should replace with undefined if value not found 1`] = `
"console.log(undefined);
"
`;

exports[`Process transform test process.xxx should inject process if a variable is not recognized 1`] = `
"var process = require(\\"process\\");
console.log(process.some());
"
`;

exports[`Process transform test process.xxx should transform process.browser 1`] = `
"console.log(true);
"
`;

exports[`Process transform test process.xxx should transform process.cwd 1`] = `
"console.log(\\"./\\");
"
`;

exports[`Process transform test process.xxx should transform process.env (not add twice) 1`] = `
"var ___env = {
  foo: \\"bar\\"
};
alert(___env);
console.log(___env);
"
`;

exports[`Process transform test process.xxx should transform process.env 1`] = `
"var ___env = {
  foo: \\"bar\\"
};
console.log(___env);
"
`;

exports[`Process transform test process.xxx should transform process.title 1`] = `
"console.log(\\"browser\\");
"
`;

exports[`Process transform test process.xxx should transform process.umask 1`] = `
"console.log(0);
"
`;

exports[`Process transform test process.xxx should transform process.version 1`] = `
"console.log(\\"v11.13.0\\");
"
`;

exports[`Process transform test process.xxx should transform process.versions 1`] = `
"console.log({});
"
`;
